src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/RazorDiagnosticsEndpoint.cs

```diff
@@ -329,17 +331,75 @@ internal virtual bool CheckIfDocumentHasRazorDiagnostic(RazorCodeDocument codeDo
             return mappedDiagnostics.ToArray();
         }
 
-        private bool TryGetOriginalDiagnosticRange(Range projectedRange, DiagnosticSeverity? severity, RazorCodeDocument codeDocument, out Range originalRange)
+        private static bool IsRudeEditDiagnostic(Diagnostic diagnostic)
         {
+            return diagnostic.Code.HasValue &&
+                diagnostic.Code.Value.IsString &&
+                diagnostic.Code.Value.String.StartsWith("ENC");
+        }
+
+        private Range RemapRudeEditRange(Range diagnosticRange, RazorCodeDocument codeDocument, SourceText sourceText)
+        {
+            // This is a rude edit diagnostic that has already been mapped to the Razor document. The mapping isn't absolutely correct though,
+            // it's based on the runtime code generation of the Razor document therefore we need to re-map the already mapped diagnostic in a
+            // semi-intelligent way.
+
+            var syntaxTree = codeDocument.GetSyntaxTree();
+            var owner = syntaxTree.GetOwner(sourceText, diagnosticRange);
+
+            switch (owner?.Kind)
+            {
+                case SyntaxKind.CSharpStatementLiteral: // Simple C# in @code block, @{ ... } etc.
+                case SyntaxKind.CSharpExpressionLiteral: // Referenced simple C# in an implicit expression @Foo((abc) => {....})
+                    // Good as is, we were able to find a known leaf-node that fully contains the diagnostic range. Therefore we can
+                    // return the diagnostic range as is.
+                    return diagnosticRange;
+
+                default:
+                    // Unsupported owner of rude diagnostic, lets map to the entirety of the diagnostic range to be sure the diagnostic can be presented
+
+                    _logger.LogInformation($"Failed to remap rude edit for SyntaxTree owner '{owner?.Kind}'.");
+
+                    var startLineIndex = diagnosticRange.Start.Line;
+                    var startLine = sourceText.Lines[startLineIndex];
```

Comment:
My experience tells me this will be trouble if we don't have guarantees that the document version that produced the diagnostics and this sourceText are the same.

Reply:
That's a great point, I should probably add some guardrails just in-case.